### PR TITLE
Flatten CustomIdentities in SharedState and Identity Refactor

### PR DESCRIPTION
**Pull request by nporter-adbe**
In Events and shared state coming from Identity, we were putting in our custom Swift type `CustomIdentity`, this format didn't match 100% with our v5 event format and also was causing issues when the data was being sent to Assurance as it was unable to deserialize the type.

Original event data:
```jsx
{
    blob = j8Odv6LonN4r3an7LhD3WZrU1bUpAkFkkiY1ncBR96t2PTI;
    lastsync = "1603311210.452279";
    locationhint = 9;
    mid = 90520654536133369506854802638635608086;
    visitoridslist =     (
        "AEPIdentity.CustomIdentity",
        "AEPIdentity.CustomIdentity"
    );
}
```

Updated event data, so now we flatten our `CustomIdentity` to a dictionary for easy reading in event data.
```jsx
{
    "visitoridslist": [
      {
        "id": "1234567",
        "authentication.state": 1,
        "id.origin": "d_cid_ic",
        "id.type": "idType1"
      }
    ],
    "mid": "47638672570484457245556507411373198888",
    "blob": "j8Odv6LonN4r3an7LhD3WZrU1bUpAkFkkiY1ncBR96t2PTI",
    "locationhint": "9"
  }
```

This PR also does some minor refactoring to reduce the number of times we need to decode a `DataEntity` when processing an `IdentityHit`.

### CODE REVIEW

File: AEPIdentity/Sources/MobileIdentities.swift [hunk 1]
```diff
@@ -127,13 +127,13 @@ struct MobileIdentities: Codable {
...
     }
-}
 
-private extension CustomIdentity {
-    /// Converts a `CustomIdentity` into a `UserID`
+    /// Converts a `CustomIdentity` dictionary representation into a `UserID`
     /// - Returns: a `UserID` where the namespace is value, value is identifier, and type is integrationCode
-    func toUserID() -> UserID? {
-        guard let type = type, let identifier = identifier, !identifier.isEmpty else { return nil }
+    private func dictToUserID(dict: [String: Any]) -> UserID? {
```
Comment (shalehaha) [inline AEPIdentity/Sources/MobileIdentities.swift:133]: can this be a static method?

**Review by shalehaha — APPROVED** [review]

File: AEPIdentity/Sources/MobileIdentities.swift (hunk 1 above)
Comment (nporter-adbe) [reply AEPIdentity/Sources/MobileIdentities.swift:133]: I suppose it could be, what's the use case?

Comment (shalehaha) [reply AEPIdentity/Sources/MobileIdentities.swift:133]: just easier to understand the behavior of the method, it won't change any instance vars

**Review by cdhoffmann — APPROVED** [review]
looks good!

**Pull request merged**